### PR TITLE
Improve readability in x/tokenfactory test

### PR DESCRIPTION
**Pull request by mattverse**
Closes: [#1845 ](https://github.com/osmosis-labs/osmosis/issues/1845)

## What is the purpose of the change

The purpose of this change is to have better readability for `admins_test.go`, where previously `TestChangeAdminDenom` had a structure where everything tested would be grouped in a single method of setup(), where now it has separate fields of testing, which consequently improves overall code readability.

## Brief Changelog

- Change `admins_test.go` to better follow tdd to improve overall readability.
- 
## Testing and Verifying

This change refactored existing tests.

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? no
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`?  no
  - How is the feature or change documented? (not applicable   /   specification (`x/<module>/spec/`)  /  [Osmosis docs repo](https://github.com/osmosis-labs/docs)   /   not documented)

### CODE REVIEW

**Review by alexanderbez — APPROVED** [review]
🧀

File: x/tokenfactory/keeper/admins_test.go [hunk 1]
```diff
@@ -185,82 +185,76 @@ func (suite *KeeperTestSuite) TestBurnDenom() {
...
-	)
-
 	for _, tc := range []struct {
-		desc  string
-		setup func()
-		valid bool
+		desc               string
+		msgChangeAdmin     func(denom string)
+		expectedAdminIndex int
+		msgMint            func(denom string)
```
Comment (ValarDragon) [inline x/tokenfactory/keeper/admins_test.go:192]: Lets make the functions returns the message type directly?

Then LGTM, ty for doing this

Comment (mattverse) [reply x/tokenfactory/keeper/admins_test.go:192]: The reason I used a separate function is bc we need to take in the changed denom as a parameter of a function to used the updated denom. 

A possible option perhaps is changing the field to become `msgChangeAdmin func(denom string) *types.MsgChangeAdmin`, where the function directly returns the message?

Comment (ValarDragon) [reply x/tokenfactory/keeper/admins_test.go:192]: yeah!

Comment (mattverse) [reply x/tokenfactory/keeper/admins_test.go]: yeah agreed that testing layers should be all together, not in separate fields!

File: x/tokenfactory/keeper/admins_test.go [hunk 2]
```diff
@@ -185,82 +185,72 @@ func (suite *KeeperTestSuite) TestBurnDenom() {
...
 
-			tc.setup()
+			// expectedAdminIndex with negative value is assumed as admin with value of ""
+			if tc.expectedAdminIndex < 0 {
+				suite.Require().Equal("", queryRes.AuthorityMetadata.Admin)
+			} else {
+				suite.Require().Equal(suite.TestAccs[tc.expectedAdminIndex].String(), queryRes.AuthorityMetadata.Admin)
+			}
+
+			if tc.msgMint != nil {
```
Comment (p0mvn) [inline x/tokenfactory/keeper/admins_test.go:250]: Let's add a comment explaining why we are also testing mint when the test spec is focused on the change admin message. I think the goal here is to make sure that we can mint after changing admin?

File: x/tokenfactory/keeper/admins_test.go [hunk 3]
```diff
@@ -185,82 +185,72 @@ func (suite *KeeperTestSuite) TestBurnDenom() {
...
-					Denom: denom,
-				})
-				suite.Require().NoError(err)
-				suite.Require().Equal("", queryRes.AuthorityMetadata.Admin)
-
-				// can't mint.
-				_, err = suite.msgServer.Mint(sdk.WrapSDKContext(suite.Ctx), types.NewMsgMint(suite.TestAccs[0].String(), sdk.NewInt64Coin(denom, 5)))
-				suite.Require().Error(err)
-				suite.Require().Equal(suite.App.BankKeeper.GetBalance(suite.Ctx, suite.TestAccs[0], denom).Amount.Int64(), addr0bal)
+			msgChangeAdmin: func(denom string) *types.MsgChangeAdmin {
```
[the remark this answers is not in the document]
Comment (p0mvn) [reply x/tokenfactory/keeper/admins_test.go:196]: Ahh I see, we can probably still initialize the variable inside the loop after calling `suite.SetupTest()`. Wdyt?

[the remark this answers is not in the document]
Comment (mattverse) [reply x/tokenfactory/keeper/admins_test.go:196]: makes sense, but we would still have to pass in the denom as a parameter right?

[the remark this answers is not in the document]
Comment (p0mvn) [reply x/tokenfactory/keeper/admins_test.go:196]: You are totally right! I didn't notice that

I'll let you decide what you think is best. This minor readability improvement is proving to be more effort than I originally expected so I think it's okay to leave this as is

[the remark this answers is not in the document]
Comment (ValarDragon) [reply x/tokenfactory/keeper/admins_test.go:196]: Also we could extend in the future to cover other perceived edge cases wrt denoms, e.g. if we were worried about creator parsing.

I;m in favor of current approach

File: x/tokenfactory/keeper/admins_test.go (hunk 2 above)
Comment (ValarDragon) [reply x/tokenfactory/keeper/admins_test.go:254]: Yeah, basically this is our "test" for the authority actually being correct

File: x/tokenfactory/keeper/admins_test.go [hunk 4]
```diff
@@ -185,82 +185,76 @@ func (suite *KeeperTestSuite) TestBurnDenom() {
...
-		addr1bal int64
-	)
-
 	for _, tc := range []struct {
-		desc  string
-		setup func()
-		valid bool
+		desc               string
+		msgChangeAdmin     func(denom string) *types.MsgChangeAdmin
+		expectedAdminIndex int
```
Comment (ValarDragon) [inline x/tokenfactory/keeper/admins_test.go:191]: Can we just make this `expectedAdmin` directly?

**Review by ValarDragon — APPROVED** [review]

File: x/tokenfactory/keeper/admins_test.go (hunk 4 above)
Comment (mattverse) [reply x/tokenfactory/keeper/admins_test.go:191]: You mean and then have them set as suite.TestAccs[0] and so on right?

Here we have the same problem where TestAccs are reseted to random accounts every iteration for each test, while the values for these fields are initialized at the very beginning

Comment (ValarDragon) [reply x/tokenfactory/keeper/admins_test.go:191]: Hrmm, not sure I get it, but not worth blocking this imo

File: x/tokenfactory/keeper/admins_test.go (hunk 2 above)
Comment (mattverse) [reply x/tokenfactory/keeper/admins_test.go:254]: added a comment here!

File: x/tokenfactory/keeper/admins_test.go [hunk 5]
```diff
@@ -185,82 +185,72 @@ func (suite *KeeperTestSuite) TestBurnDenom() {
...
+			// expectedAdminIndex with negative value is assumed as admin with value of ""
+			if tc.expectedAdminIndex < 0 {
+				suite.Require().Equal("", queryRes.AuthorityMetadata.Admin)
+			} else {
+				suite.Require().Equal(suite.TestAccs[tc.expectedAdminIndex].String(), queryRes.AuthorityMetadata.Admin)
+			}
+
+			if tc.msgMint != nil {
+				_, err := suite.msgServer.Mint(sdk.WrapSDKContext(suite.Ctx), tc.msgMint(testDenom))
+				suite.Require().NoError(err)
```
[the remark this answers is not in the document]
Comment (mattverse) [reply x/tokenfactory/keeper/admins_test.go:256]: The sencond test case would be the case!

**Pull request merged**